File: apps/desktop/src/renderer/features/sidebar/characterPackV3.ts
/**
 * 模块用途：定义并严格校验十四行 CharacterPack V3，同时把旧角色包映射到统一动作槽位。
 * 模块边界：只处理清单数据，不加载图像，也不决定动作时长和状态优先级。
 */
import type { PetDragDirection } from "./petDragDirection";

export interface CharacterClipV3 {
  row: number;
  mirrorX?: boolean;
}

export interface CharacterAtlasV3 {
  columns: 6;
  rows: 7 | 14;
  cellWidth: 192;
  cellHeight: 208;
  renderWidth: 88;
  renderHeight: 96;
}

export interface CharacterClipsV3 {
  idle: { base: CharacterClipV3; blink: CharacterClipV3 };
  awaken: CharacterClipV3;
  dragging: Record<PetDragDirection, CharacterClipV3>;
  thinking: CharacterClipV3;
  working: CharacterClipV3;
  waiting: CharacterClipV3;
  reminding: CharacterClipV3;
  complete: CharacterClipV3;
  error: CharacterClipV3;
  sleeping: CharacterClipV3;
}

export interface CharacterPackManifestV3 {
  schemaVersion: 3;
  id: string;
  displayName: string;
  version: string;
  atlas: CharacterAtlasV3;
  clips: CharacterClipsV3;
}

interface LegacyClip {
  row: number;
  mirrorX?: boolean;
}

interface LegacyManifest {
  id: string;
  displayName: string;
  version: string;
  atlas: Omit<CharacterAtlasV3, "rows"> & { rows: 7 };
  clips: {
    idle: LegacyClip;
    awaken: LegacyClip;
    dragging: Record<PetDragDirection, LegacyClip>;
    working: LegacyClip;
    complete: LegacyClip;
  };
}

const ROOT_KEYS = ["schemaVersion", "id", "displayName", "version", "atlas", "clips"];
const CLIP_ROWS = {
  awaken: 2,
  thinking: 7,
  working: 8,
  waiting: 9,
  reminding: 10,
  complete: 11,
  error: 12,
  sleeping: 13
} as const;

export function validateCharacterPackV3(value: Record<string, unknown>): CharacterPackManifestV3 {
  exactKeys(value, ROOT_KEYS, "角色包");
  validateIdentity(value);
  validateAtlas(value.atlas);
  const clips = record(value.clips, "动作");
  exactKeys(clips, ["idle", "awaken", "dragging", "thinking", "working", "waiting",
    "reminding", "complete", "error", "sleeping"], "动作");
  validateIdle(clips.idle);
  validateDragging(clips.dragging);
  for (const [name, expectedRow] of Object.entries(CLIP_ROWS)) {
    validateClip(clips[name], name, expectedRow);
  }
  return value as unknown as CharacterPackManifestV3;
}

export function upgradeLegacyCharacterPack(legacy: LegacyManifest): CharacterPackManifestV3 {
  const old = legacy.clips;
  return {
    schemaVersion: 3,
    id: legacy.id,
    displayName: legacy.displayName,
    version: legacy.version,
    atlas: legacy.atlas,
    clips: {
      idle: { base: copy(old.idle), blink: copy(old.idle) },
      awaken: copy(old.awaken),
      dragging: {
        down: copy(old.dragging.down), up: copy(old.dragging.up),
        right: copy(old.dragging.right), left: copy(old.dragging.left)
      },
      thinking: copy(old.working),
      working: copy(old.working),
      waiting: copy(old.idle),
      reminding: copy(old.awaken),
      complete: copy(old.complete),
      error: copy(old.idle),
      sleeping: copy(old.idle)
    }
  };
}

function validateIdentity(value: Record<string, unknown>) {
  if (typeof value.id !== "string" || !/^[a-z0-9][a-z0-9-]*$/.test(value.id)) {
    throw new Error("角色 ID 无效");
  }
  if (typeof value.displayName !== "string" || !value.displayName.trim()) throw new Error("角色名称无效");
  if (typeof value.version !== "string" || !value.version.trim()) throw new Error("角色版本无效");
}

function validateAtlas(value: unknown) {
  const atlas = record(value, "图集");
  exactKeys(atlas, ["columns", "rows", "cellWidth", "cellHeight", "renderWidth", "renderHeight"], "图集");
  const expected = { columns: 6, rows: 14, cellWidth: 192, cellHeight: 208, renderWidth: 88, renderHeight: 96 };
  for (const [key, expectedValue] of Object.entries(expected)) {
    if (atlas[key] !== expectedValue) throw new Error(`图集尺寸字段 ${key} 无效`);
  }
}

function validateIdle(value: unknown) {
  const idle = record(value, "idle");
  exactKeys(idle, ["base", "blink"], "idle");
  validateClip(idle.base, "idle.base", 0);
  validateClip(idle.blink, "idle.blink", 1);
}

function validateDragging(value: unknown) {
  const dragging = record(value, "dragging");
  exactKeys(dragging, ["down", "up", "right", "left"], "dragging");
  validateClip(dragging.down, "dragging.down", 3);
  validateClip(dragging.up, "dragging.up", 4);
  validateClip(dragging.right, "dragging.right", 5);
  validateClip(dragging.left, "dragging.left", 6);
}

function validateClip(value: unknown, name: string, expectedRow: number) {
  const clip = record(value, name);
  exactKeys(clip, ["row"], name);
  if (clip.row !== expectedRow) throw new Error(`${name} 行号必须为 ${expectedRow}`);
}

function copy(clip: LegacyClip): CharacterClipV3 {
  return { row: clip.row, ...(clip.mirrorX === undefined ? {} : { mirrorX: clip.mirrorX }) };
}

function record(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new Error(`${label}必须是对象`);
  return value as Record<string, unknown>;
}

function exactKeys(value: Record<string, unknown>, keys: string[], label: string) {
  const actual = Object.keys(value).sort();
  const expected = [...keys].sort();
  if (actual.length !== expected.length || actual.some((key, index) => key !== expected[index])) {
    throw new Error(`${label}包含未知字段`);
  }
}

File: scripts/pets/character_pack_v3.py
"""角色工厂 V3 的固定尺寸、动作行和清单合同。"""

from __future__ import annotations

import re
from typing import Any

FRAME_COUNT = 6
ROW_COUNT = 14
CELL_WIDTH, CELL_HEIGHT = 192, 208
ATLAS_WIDTH, ATLAS_HEIGHT = 1152, 2912
RENDER_WIDTH, RENDER_HEIGHT = 88, 96

STATE_SPECS = (
    ("idle-base", 0, 4800, True),
    ("idle-blink", 1, 360, False),
    ("awaken", 2, 840, False),
    ("drag-down", 3, 600, True),
    ("drag-up", 4, 600, True),
    ("drag-right", 5, 600, True),
    ("drag-left", 6, 600, True),
    ("thinking", 7, 2800, True),
    ("working", 8, 1400, True),
    ("waiting", 9, 3200, True),
    ("reminding", 10, 800, False),
    ("complete", 11, 900, False),
    ("error", 12, 900, False),
    ("sleeping", 13, 6000, True),
)
STATE_NAMES = tuple(spec[0] for spec in STATE_SPECS)


def build_manifest(character_id: str, display_name: str, version: str = "3.0.0") -> dict[str, Any]:
    """创建只含身份、图集规格和动作行映射的 V3 清单。"""
    return {
        "schemaVersion": 3,
        "id": character_id,
        "displayName": display_name,
        "version": version,
        "atlas": {
            "columns": FRAME_COUNT,
            "rows": ROW_COUNT,
            "cellWidth": CELL_WIDTH,
            "cellHeight": CELL_HEIGHT,
            "renderWidth": RENDER_WIDTH,
            "renderHeight": RENDER_HEIGHT,
        },
        "clips": {
            "idle": {"base": {"row": 0}, "blink": {"row": 1}},
            "awaken": {"row": 2},
            "dragging": {
                "down": {"row": 3}, "up": {"row": 4},
                "right": {"row": 5}, "left": {"row": 6},
            },
            "thinking": {"row": 7},
            "working": {"row": 8},
            "waiting": {"row": 9},
            "reminding": {"row": 10},
            "complete": {"row": 11},
            "error": {"row": 12},
            "sleeping": {"row": 13},
        },
    }


def validate_manifest(manifest: object) -> dict[str, Any]:
    """严格拒绝未知字段和错误行号。"""
    value = _record(manifest, "角色包")
    _exact(value, {"schemaVersion", "id", "displayName", "version", "atlas", "clips"}, "角色包")
    if value.get("schemaVersion") != 3:
        raise ValueError("角色包 schemaVersion 必须为 3")
    if not isinstance(value.get("id"), str) or not re.fullmatch(r"[a-z0-9][a-z0-9-]*", value["id"]):
        raise ValueError("角色 ID 无效")
    if not isinstance(value.get("displayName"), str) or not value["displayName"].strip():
        raise ValueError("角色名称无效")
    version = value.get("version")
    if not isinstance(version, str) or not version.strip():
        raise ValueError("角色版本无效")
    expected = build_manifest(value["id"], value["displayName"], version)
    if value.get("atlas") != expected["atlas"]:
        raise ValueError("图集尺寸合同无效")
    if value.get("clips") != expected["clips"]:
        raise ValueError("十四行动作映射无效或包含未知字段")
    return value


def _record(value: object, label: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError(f"{label}必须是对象")
    return value


def _exact(value: dict[str, Any], keys: set[str], label: str) -> None:
    if set(value) != keys:
        raise ValueError(f"{label}包含未知字段")
